File: socket_programming_samples/chatroom_client.py
###### IN PROGRESS ######

from concurrent.futures import thread
from sre_constants import FAILURE, SUCCESS
import threading
from signal import SIGINT, signal
import socket
import struct

CLIENT_CAPACITY = 100
BUFFER_SIZE = 2048
NAME_LEN = 32

flag = 0
sockfd = 0
name = ""

class Thread(threading.Thread):
    def run(self):
        pass

def catch_ctr_c_and_exit():
    flag = 1

def string_overwrite():
    print("\r%s", "> ")
    fflush(stdout)


def string_trim(arr, length):

    for i in range(length):
        if(arr[i] == '\n'):
            arr[i] = '\0'
            break

def send_message_handler():
    message = ""
    buffer = ""

    while 1:
        string_overwrite()
    
        string_trim(message, BUFFER_SIZE)
        recieve = socket.recv_fds(sockfd, message, BUFFER_SIZE, 0)

        if "exit" in buffer:
            pass

def recv_message_handler():
    message = ""

    while 1:
        recieve = socket.recv_fds(sockfd, message, BUFFER_SIZE, 0)

        if recieve > 0:
            print(message)
            string_overwrite()
        elif recieve == 0:
            break

        message = ""
        
def main(argc, argv):
    if argc != 2:
        print("Usage: %s <port>\n", argv[0])
        return -1

    IP = "127.0.0.1"
    PORT = argv[0]
    ADDRESS = (IP, PORT)
    SIZE = 1024
    FORMAT = "utf-8"
    
    name = input("Enter your name: ")

    signal(SIGINT, catch_ctr_c_and_exit)
    string_trim(name, len(name))

    if len(name) > NAME_LEN - 1 or len(name) < 2:
        print(f"ERROR: Invalid Name. Must be between 2-{NAME_LEN} characters \n")

    err = socket.connect(ADDRESS)

    if err == -1:
        print("ERROR: connect\n")
        return -1

    # send the name
    socket.send_fds(sockfd, name, NAME_LEN, 0)

    print("~~~~~Welcome to the Chat Room!~~~~~\n")

    send_msg_thread = Thread(name="Thread-Send-{}".format(name))

    if not send_msg_thread.is_alive:
        print("ERROR: send_msg_thread failed to run\n")
        return FAILURE

    recv_msg_thread = Thread(name="Thread-Recieve-{}".format)

    if not recv_msg_thread.is_alive:
        print("ERROR: recv_msg_thread failed to run\n")
        return FAILURE
    
    while 1:
        if flag:
            print("\nGoodbye\n")
            break

        socket.close(sockfd)

        return SUCCESS

if __name__ == "__main__":
    main()

File: chatroom_client.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <signal.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <pthread.h>

#define BUFFER_SIZE 2048
#define NAME_LEN 32

// Global variables
volatile sig_atomic_t flag = 0;
int sockfd = 0;
char name[32];

void str_overwrite_stdout()
{
    printf("%s", "> ");
    fflush(stdout);
}

void string_trim(char *msg, int length)
{
    int i;
    for (i = 0; i < length; i++)
    {
        if (msg[i] == '\n')
        {
            msg[i] = '\0';
            break;
        }
    }
}

void catch_ctrl_c_and_exit()
{
    flag = 1;
}

void send_msg_handler()
{
    char buffer[BUFFER_SIZE] = {};
    char message[BUFFER_SIZE + NAME_LEN] = {};

    while (1)
    {
        str_overwrite_stdout();
        fgets(message, BUFFER_SIZE, stdin);
        string_trim(message, BUFFER_SIZE);

        if (strcmp(message, "exit") == 0)
        {
            break;
        }
        else
        {
            sprintf(buffer, "%s: %s\n", name, message);
            send(sockfd, buffer, strlen(buffer), 0);
        }

        bzero(buffer, BUFFER_SIZE);
        bzero(message, BUFFER_SIZE + NAME_LEN);
    }
    catch_ctrl_c_and_exit();
}

void recv_msg_handler()
{
    char message[BUFFER_SIZE] = {};

    while (1)
    {
        int receive = recv(sockfd, message, BUFFER_SIZE, 0);
        if (receive > 0)
        {
            printf("%s", message);
            str_overwrite_stdout();
        }
        else if (receive == 0)
        {
            break;
        }
        bzero(message, BUFFER_SIZE);
    }
}

int main(int argc, char **argv)
{
    if (argc != 2)
    {
        printf("Usage: %s <port>\n", argv[0]);
        return EXIT_FAILURE;
    }

    char *ip = "127.0.0.1";
    int port = atoi(argv[1]);

    signal(SIGINT, catch_ctrl_c_and_exit);

    printf("Enter your name: ");
    fgets(name, 32, stdin);
    string_trim(name, strlen(name));

    if (strlen(name) > 32 || strlen(name) < 2)
    {
        printf("Name must be less than 30 and more than 2 characters.\n");
        return EXIT_FAILURE;
    }

    struct sockaddr_in server_address;

    /* Socket settings */
    sockfd = socket(AF_INET, SOCK_STREAM, 0);
    server_address.sin_family = AF_INET;
    server_address.sin_addr.s_addr = inet_addr(ip);
    server_address.sin_port = htons(port);

    // Connect to Server
    int err = connect(sockfd, (struct sockaddr *)&server_address, sizeof(server_address));
    if (err == -1)
    {
        printf("ERROR: connect\n");
        return EXIT_FAILURE;
    }

    // Send name
    send(sockfd, name, 32, 0);

    printf("~~~~~ Welcome to the Chat Room ~~~~~\n");

    pthread_t send_msg_thread;
    if (pthread_create(&send_msg_thread, NULL, (void *)send_msg_handler, NULL) != 0)
    {
        printf("ERROR: pthread\n");
        return EXIT_FAILURE;
    }

    pthread_t recv_msg_thread;
    if (pthread_create(&recv_msg_thread, NULL, (void *)recv_msg_handler, NULL) != 0)
    {
        printf("ERROR: pthread\n");
        return EXIT_FAILURE;
    }

    while (1)
    {
        if (flag)
        {
            printf("\nSalutations %s...\n", name);
            break;
        }
    }

    close(sockfd);

    return EXIT_SUCCESS;
}